### PR TITLE
Fix bug where the configuration of the hbaseRootdir at the role level is ignored

### PR DESCRIPTION
**Pull request by maltesander**
# Description

The problem was defaulting in the product config callbacks like:

```
 result.insert( 
     HBASE_ROOTDIR.to_string(), 
     Some( 
         self.hbase_rootdir 
             .as_deref() 
             .unwrap_or(HBASE_ROOT_DIR_DEFAULT) 
             .to_string(), 
     ), 
 ); 
```

If the `hbaseRootdir` was only set on the role, we defaulted the rolegroup to `HBASE_ROOT_DIR_DEFAULT` ("/hbase"), which during merging overrode the value set on the role. The solution here is to remove the defaulting and it works.
One problem remained, if `hbaseRootdir` was not set at all, there was no default set anymore. The product config did not add the default due to the field being marked as not required.
In order to keep the defaulting behavior as is and not mess with existing defaulted installation, the default for `hbase.root.dir` in the product config was changed to "/hbase" and the field marked as required (on all roles).
The docs state though:
```
The directory shared by region servers and into which HBase persists. The URL should be 'fully-qualified' to include the filesystem scheme. For example, to specify the HDFS directory '/hbase' where the HDFS instance's namenode is running at namenode.example.org on port 9000, set this value to: hdfs://namenode.example.org:9000/hbase. By default, we write to whatever ${hbase.tmp.dir} is set too -- usually /tmp -- so change this configuration or else all data will be lost on machine restart.
```

So i think we can only use it on the regionservers? Ping @lfrancke 

Edit: Lars would bet its at least required in the masters, so i would keep it as is (required in all roles in the product config)

## Definition of Done Checklist

- Not all of these items are applicable to all PRs, the author should update this template to only leave the boxes in that are relevant
- Please make sure all these things are done and tick the boxes

```[tasklist]
# Author
- [ ] Changes are OpenShift compatible
- [ ] CRD changes approved
- [ ] CRD documentation for all fields, following the [style guide](https://docs.stackable.tech/home/nightly/contributor/docs/style-guide).
- [ ] Helm chart can be installed and deployed operator works
- [ ] Integration tests passed (for non trivial changes)
- [ ] Changes need to be "offline" compatible
```

```[tasklist]
# Reviewer
- [ ] (Integration-)Test cases added
- [x] Changelog updated
- [x] Cargo.toml only contains references to git tags (not specific commits or branches)
```

```[tasklist]
# Acceptance
- [ ] Feature Tracker has been updated
- [ ] Proper release label has been added
- [ ] [Roadmap](https://github.com/orgs/stackabletech/projects/25/views/1) has been updated
```

### CODE REVIEW

File: deploy/config-spec/properties.yaml [hunk 1]
```diff
@@ -213,14 +213,14 @@ properties:
...
       roles:
         - name: "master"
-          required: false
+          required: true # TODO: this could be false?
         - name: "regionserver"
-          required: false
+          required: true
         - name: "restserver"
-          required: false
+          required: true # TODO: this could be false?
```
Comment (siegfriedweber) [inline deploy/config-spec/properties.yaml:223]: Only the `hbase.rootdir` is effective which is configured at the master. The other ones are ignored.

**Review by siegfriedweber — CHANGES_REQUESTED** [review]

File: rust/crd/src/lib.rs [hunk 2]
```diff
@@ -473,15 +473,7 @@ impl Configuration for HbaseConfigFragment {
                     HBASE_UNSAFE_REGIONSERVER_HOSTNAME_DISABLE_MASTER_REVERSEDNS.to_string(),
                     Some("true".to_string()),
                 );
-                result.insert(
-                    HBASE_ROOTDIR.to_string(),
-                    Some(
-                        self.hbase_rootdir
-                            .as_deref()
-                            .unwrap_or(HBASE_ROOT_DIR_DEFAULT)
```
[the remark this answers is not in the document]
Comment (maltesander) [reply rust/crd/src/lib.rs:481]: good point, done https://github.com/stackabletech/hbase-operator/pull/584/commits/28b556c67c3a633be8f1b6d088783471a7d412fc

File: deploy/config-spec/properties.yaml (hunk 1 above)
Comment (maltesander) [reply deploy/config-spec/properties.yaml]: improved/corrected comments https://github.com/stackabletech/hbase-operator/pull/584/commits/1095505a4371acfaa0c2b9adbabf825c3f90ee6a and keep restserver and and regionserver true for compatibility reasons.

**Review by siegfriedweber — APPROVED** [review]
LGTM

**Comment by lfrancke** [comment]
I'm a bit confused by the comment. That seems to indicate that it's not required by the RegionServer?

**Comment by siegfriedweber** [comment]
> I'm a bit confused by the comment. That seems to indicate that it's not required by the RegionServer?

The master apparently tells the RegionServers where the root dir is located. I have not checked this assumption in the code but I tested it:

If I set different root dirs for the roles ...

```diff
$ git diff
diff --git a/tests/templates/kuttl/smoke/30-install-hbase.yaml.j2 b/tests/templates/kuttl/smoke/30-install-hbase.yaml.j2
index ac1bed8..ea5d883 100644
--- a/tests/templates/kuttl/smoke/30-install-hbase.yaml.j2
+++ b/tests/templates/kuttl/smoke/30-install-hbase.yaml.j2
@@ -21,6 +21,7 @@ spec:
 {% endif %}
   masters:
     config:
+      hbaseRootdir: /rtmaster
       logging:
         enableVectorAgent: {{ lookup('env', 'VECTOR_AGGREGATOR') | length > 0 }}
     roleGroups:
@@ -32,6 +33,7 @@ spec:
         replicas: 2
   regionServers:
     config:
+      hbaseRootdir: /rtregion
       logging:
         enableVectorAgent: {{ lookup('env', 'VECTOR_AGGREGATOR') | length > 0 }}
     roleGroups:
@@ -43,6 +45,7 @@ spec:
         replicas: 2
   restServers:
     config:
+      hbaseRootdir: /rtrest
       logging:
         enableVectorAgent: {{ lookup('env', 'VECTOR_AGGREGATOR') | length > 0 }}
     roleGroups:
```

... then the one set for the master is effective:

```shell
$ bin/hdfs dfs -fs hdfs://test-hdfs/ -ls /
Found 1 items
drwxr-xr-x   - stackable supergroup          0 2024-11-15 11:00 /rtmaster
$ bin/hdfs dfs -fs hdfs://test-hdfs/ -ls /rtmaster
Found 12 items
drwxr-xr-x   - stackable supergroup          0 2024-11-15 11:00 /rtmaster/.hbck
drwxr-xr-x   - stackable supergroup          0 2024-11-15 11:00 /rtmaster/.tmp
drwxr-xr-x   - stackable supergroup          0 2024-11-15 11:00 /rtmaster/MasterData
drwxr-xr-x   - stackable supergroup          0 2024-11-15 11:00 /rtmaster/WALs
drwxr-xr-x   - stackable supergroup          0 2024-11-15 11:10 /rtmaster/archive
drwxr-xr-x   - stackable supergroup          0 2024-11-15 11:00 /rtmaster/corrupt
drwxr-xr-x   - stackable supergroup          0 2024-11-15 11:01 /rtmaster/data
-rw-r--r--   3 stackable supergroup         42 2024-11-15 11:00 /rtmaster/hbase.id
-rw-r--r--   3 stackable supergroup          7 2024-11-15 11:00 /rtmaster/hbase.version
drwxr-xr-x   - stackable supergroup          0 2024-11-15 11:00 /rtmaster/mobdir
drwxr-xr-x   - stackable supergroup          0 2024-11-15 11:00 /rtmaster/oldWALs
drwx--x--x   - stackable supergroup          0 2024-11-15 11:00 /rtmaster/staging
```

**Comment by maltesander** [comment]
To add to this, we (or I?) decided to still set the `hbaseRootdir` on the region and rest server (even if not required) in order to keep existing "functionality" or at least avoid restarts in exitsting installations due to config changes.

**Pull request merged**